pkg/controller/elasticsearch/driver/downscale.go

```diff
@@ -77,11 +77,24 @@ func calculateDownscales(expectedStatefulSets sset.StatefulSetList, actualStatef
 		}
 		if expectedReplicas == 0 || // removal
 			expectedReplicas < actualReplicas { // downscale
+			if canDownscale, reason := checkDownscaleInvariants(state, actualSset); !canDownscale {
+				ssetLogger(actualSset).V(1).Info("Cannot downscale StatefulSet", "reason", reason)
+				continue
+			}
+
+			toDelete := actualReplicas - expectedReplicas
+			if label.IsMasterNodeSet(actualSset) {
```

Comment:
Why is it not part of `checkDownscaleInvariants `? Also we should not downscale more than one master for the entire cluster (not per StatefulSet).

Reply:
Edit: maybe more part of `getMaxNodesToRemove`

Reply:
I feel like checkDownscaleInvariants and getMaxNodesToRemove could maybe be merged in a single function that returns the number of nodes that can be downscaled (could be 0), and the reason why (can be nil).

Reply:
In the interest of time, could this be in a separate, refactoring PR?

Reply:
I don't know if I would include `recordRemoval ` in that function. In order to keep a clear separation between the function that does the check, and the function that applies the side-effects.
Totally fine to do in a follow-up PR.

Reply:
Sounds good.